designer/server/src/main/scala/pl/touk/nussknacker/ui/util/ApiAdapter.scala

```diff
@@ -0,0 +1,6 @@
+package pl.touk.nussknacker.ui.util
+
+trait ApiAdapter[D] {
```

Comment:
I wonder if we shouldn't have here 3 generic params:
* previous version
* current version
* next version

Obviously, we cannot assume there will be always possible to lift or downgrade the version so each method should rather have declarations like that:

```scala
def liftVersion: CURRENT_VERSION => Option[NEXT_VERSION]
def downgradeVersion: CURRENT_VERSION => Option[PREV_VERSION]
```

Reply:
Yeah typesafety would be great. But note that composition of such adapters becomes more problematic. And it is easy to end up in the world full of `Foo[D, Any, Any, Any, Any]`. I'll mark it for the future improvement.